### PR TITLE
lsat: introduce LSAT related utilities

### PR DESCRIPTION
**Pull request by wpaulino**
We introduce a new package: `lsat`, which aims to provide utilities that will serve useful in the context of LSAT creation and verification for LSAT-enabled services.

### CODE REVIEW

File: lsat/caveat.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	for condition, caveats := range relevantCaveats {
+		satisfier := caveatSatisfiers[condition]
+
+		// Since it's possible for a chain of caveat to exist for the
+		// same condition as a way to demote privileges, we'll ensure
+		// each one satisfies its previous.
+		if len(caveats) > 1 {
+			for i, caveat := range caveats[1:] {
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply lsat/caveat.go]: Done.

File: lsat/satisfier.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+	// SatisfyPrevious ensures a caveat is in accordance with a previous one
+	// with the same condition. This is needed since caveats of the same
+	// condition can be used multiple times as long as they enforce more
+	// permissions than the previous.
+	//
+	// For example, we have a caveat that only allows us to use an LSAT for
+	// 7 more days. We can add another caveat that only allows for 3 more
+	// days of use and lend it to another party.
+	SatisfyPrevious func(Caveat, Caveat) error
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply lsat/satisfier.go]: Added var names to the interface method and flipped them.

File: lsat/identifier.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+	switch id.Version {
+	// A version 0 identifier consists of its linked payment hash, followed
+	// by the token ID.
+	case 0:
+		if _, err := w.Write(id.PaymentHash[:]); err != nil {
+			return err
+		}
+		_, err := w.Write(id.TokenID[:])
+		return err
```
Comment (guggero) [inline lsat/identifier.go]: nit: there should be a new line before the next case.

File: lsat/identifier.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		var tokenID [TokenIDSize]byte
+		if _, err := r.Read(tokenID[:]); err != nil {
+			return nil, err
+		}
+
+		return &Identifier{
+			Version:     version,
+			PaymentHash: paymentHash,
+			TokenID:     tokenID,
+		}, nil
```
Comment (guggero) [inline lsat/identifier.go]: nit: newline

File: lsat/satisfier.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+			return nil
+		},
+		SatisfyFinal: func(c Caveat) error {
+			services, err := decodeServicesCaveatValue(c.Value)
+			if err != nil {
+				return err
+			}
+			for _, service := range services {
+				if service.Name == targetService {
```
Comment (guggero) [inline lsat/satisfier.go]: Maybe add a TODO here that the tier verification also should be done once we have more than one tier.

**Pull request merged**